File: Sources/Celler.App.Web/Game/Client/Managers/SessionManager.ts
module Celler {

    export class SessionManager {
        game: Phaser.Game;
        id: string;

        constructor( game: Phaser.Game ) {
            this.game = game;

            app.server.getSession().done( ( sesion: SessionModel ) => {
                this.fromModel( sesion );
            } );
            app.server.onFoodAdded.add( this.onFoodAdded, this );
            app.server.onFoodRemoved.add( this.onFoodRemoved, this );
            app.server.onFoodsUpdated.add( this.onFoodsUpdated, this );
        }

        private fromModel( model: SessionModel ) {
            this.id = model.Id;
            this.createHomes( model.Homes );
            this.createCells( model.Cells );
            this.createSights( model.Sights );
            this.createFoods( model.Foods );
        }

        private createHomes( arr: HomeModel[] ) {
            arr.map( model => { this.game.add.existing( new Home( this.game, model ) ); } );
        }

        private createCells( arr: CellModel[] ) {
            arr.map( model => { this.game.add.existing( new Cell( this.game, model ) ); } );
        }

        private createSights( arr: SightModel[] ) {
            arr.map( model => { this.game.add.existing( new Sight( this.game, model ) ); } );
        }

        private createFoods( arr: FoodModel[] ) {
            arr.map( model => this.addFood( model ) );
        }


        foods: { [ id: string ]: Food; } = {};

        private addFood( model: FoodModel ) {
            var food = new Food( this.game, model );
            this.foods[ food.id ] = food;
            this.game.add.existing( food );
        }

        private onFoodAdded( model: FoodModel ) {
            this.addFood( model );
        }

        private onFoodRemoved( id: string ) {
            var food = this.foods[ id ];
            this.game.world.remove( food );
            food.destroy( true );
        }

        private onFoodsUpdated( models: FoodModel[] ) {
            
            models.forEach( model => {
                this.updateFood(this.foods[model.Base.Id], model);
            } );
        }

        private updateFood( food: Food, model: FoodModel ) {
            food.resize( model.Base.Size );
        }
    }
}

File: Sources/Celler.App.Web/Game/Server/Managers/SessionManager.cs
// Celler (c) 2015 Krokodev
// Celler.App.Web
// SessionManager.cs

using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using Celler.App.Web.Game.Server.Clients;
using Celler.App.Web.Game.Server.Entities.Abstract;
using Celler.App.Web.Game.Server.Entities.Enums;
using Celler.App.Web.Game.Server.Entities.GameObjects;
using Celler.App.Web.Game.Server.Entities.Interfaces;
using Celler.App.Web.Game.Server.Entities.Structs;
using Celler.App.Web.Game.Server.Logic;
using Celler.App.Web.Game.Server.Models;

namespace Celler.App.Web.Game.Server.Managers
{
    public class SessionManager : Entity< SessionModel >, IFoodManager, IBodyManager, ICellManager, IHomeManager,
        ISightManager
    {
        #region Constructor

        public SessionManager( IGameClient clients )
        {
            _id = Guid.NewGuid().ToString();
            _cells = new List< Cell >();
            _homes = new List< Home >();
            _sights = new List< Sight >();
            _foods = new List< Food >();
            _clients = clients;
        }

        #endregion


        #region IEntity

        protected override SessionModel ToModel()
        {
            return new SessionModel {
                Id = _id,
                Cells = _cells.Select( o => o.IModelled.Model ).ToArray(),
                Homes = _homes.Select( o => o.IModelled.Model ).ToArray(),
                Sights = _sights.Select( o => o.IModelled.Model ).ToArray(),
                Foods = _foods.Select( o => o.IModelled.Model ).ToArray()
            };
        }

        #endregion


        #region ICellManager

        public ICellManager ICellManager
        {
            get { return this; }
        }

        Cell ICellManager.AddCell( Suit suit, Point position, double size )
        {
            var cell = new Cell( suit, position, size );
            _cells.Add( cell );
            return cell;
        }

        void ICellManager.MoveCell( string id, PointModel position )
        {
            _cells.First( c => c.IIdentifiable.Id == id ).IBody.Position = new Point( position );
            _clients.CellMoved( id, position );
        }

        #endregion


        #region IFoodManager

        public IFoodManager IFoodManager
        {
            get { return this; }
        }

        Food IFoodManager.AddFood( Suit suit, Point position, double size, DateTime time, double maxValue, double frequancy )
        {
            var food = new Food( suit, position, size, time, maxValue, frequancy);
            _foods.Add( food );
            _clients.FoodAdded( food.IModelled.Model );
            return food;
        }

        void IFoodManager.RemoveFood( Food food )
        {
            _foods.RemoveAll( f => f.IIdentifiable.Id == food.IIdentifiable.Id );
            _clients.FoodRemoved( food.IIdentifiable.Id );
        }

        int IFoodManager.GetFoodCount()
        {
            return _foods.Count();
        }

        void IFoodManager.UpdateFoods( Action< Food > action )
        {
            _foods.ForEach( action );

            _clients.FoodsUpdated( _foods.Select( f=>f.IModelled.Model ).ToArray() );
        }

        #endregion


        #region IBodyManager

        public IBodyManager IBodyManager
        {
            get { return this; }
        }

        IList< IBody > IBodyManager.GetBodies()
        {
            return _homes
                .Concat< IBody >( _cells )
                .Concat< IBody >( _foods )
                .ToList();
        }

        #endregion


        #region IHomeManager

        public IHomeManager IHomeManager
        {
            get { return this; }
        }

        Home IHomeManager.AddHome( Suit suit, Point position, double size )
        {
            var obj = new Home( suit, position, size );
            _homes.Add( obj );
            return obj;
        }

        #endregion


        #region ISightManager

        public ISightManager ISightManager
        {
            get { return this; }
        }

        Sight ISightManager.AddSight( Suit suit, Point position, double size )
        {
            var obj = new Sight( suit, position, size );
            _sights.Add( obj );
            return obj;
        }

        void ISightManager.MoveSight( string id, PointModel position )
        {
            _sights.First( c => c.IIdentifiable.Id == id ).IBody.Position = new Point( position );
            _clients.SightMoved( id, position );
        }

        #endregion


        #region Fields

        private readonly string _id;
        private readonly List< Cell > _cells;
        private readonly List< Home > _homes;
        private readonly List< Sight > _sights;
        private readonly List< Food > _foods;
        private readonly IGameClient _clients;

        #endregion
    }
}